### ISSUE TITLE
const functions can assert! but cannot assert_eq!

### ISSUE THREAD

**Issue by oconnor663** [rust-lang/rust#119826]
This is probably a well-known limitation, but I couldn't find any other issues for it with a quick search. Apologies if it's a duplicate.

```rust
// compiles
pub const fn foo() {
    assert!(1 == 1);
}

// does not compile
pub const fn bar() {
    assert_eq!(1, 1);
}
```

```
error[E0015]: cannot call non-const fn `core::panicking::assert_failed::<i32, i32>` in constant functions
 --> src/lib.rs:6:5
  |
6 |     assert_eq!(1, 1);
  |     ^^^^^^^^^^^^^^^^
  |
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b0862852c785570e2dfe8d41a3767729)

Can `core::panicking::assert_failed` be made `const`?

**Comment by krtab** [comment]
Hi!

I believe this is due to the fact that `assert_eq` (and other `assert_*` macros) end up calling [core::panicking::assert_fail_inner](https://github.com/rust-lang/rust/blob/3a6bf351a39b35a662022c151621b932b4ae0bc8/library/core/src/panicking.rs#L324). Which in turn uses formatting macros (`println!` with a `{:?}` placeholder) that are currently not usable in a const context which is [known](https://blog.rust-lang.org/2021/12/02/Rust-1.57.0.html#panic-in-const-contexts) but I couldn't find any issue tracking progress on this front.

**Comment by saethlin** [comment]
Yeah, I think this is blocked on const traits. `assert_eq!` can only be used when the `Debug` impls or the tail are guaranteed to work in const, and we don't have any way to express that yet.

**Comment by Randl** [comment]
We could use `const_panic`  meanwhile?
https://github.com/rust-lang/rust/blob/2783fc43fd13a669f314742890acd64200ad0bbf/library/core/src/panic.rs#L193-L203

**Comment by oli-obk** [comment]
We don't need hacks, we can start marking Debug as const_trait and making impls const as we go. Then we need to figure out how to make more Formatter methods const
